pkg/relayer/session/claim.go

```diff
@@ -15,132 +15,167 @@ import (
-		Str("hash", fmt.Sprintf("%x", createClaimWindowStartBlock.Hash())).
+		Str("hash", fmt.Sprintf("%x", createClaimsWindowStartBlock.Hash())).
 		Msg("waiting & blocking for earliest claim creation height for this supplier")
 
 	_ = rs.waitForBlock(ctx, earliestCreateClaimHeight)
+
+	return <-claimsFlushedCh
 }
 
-// newMapClaimSessionFn returns a new MapFn that creates a claim for the given
-// session. Any session which encounters an error while creating a claim is sent
-// on the failedCreateClaimSessions channel.
-func (rs *relayerSessionsManager) newMapClaimSessionFn(
-	failedCreateClaimSessionsPublishCh chan<- relayer.SessionTree,
-) channel.MapFn[relayer.SessionTree, either.SessionTree] {
+// newMapClaimSessionsFn returns a new MapFn that creates claims for the given
+// session number. Any session which encounters an error while creating a claim
+// is sent on the failedCreateClaimSessions channel.
+func (rs *relayerSessionsManager) newMapClaimSessionsFn(
+	failedCreateClaimsSessionsPublishCh chan<- []relayer.SessionTree,
+) channel.MapFn[[]relayer.SessionTree, either.SessionTrees] {
 	return func(
 		ctx context.Context,
-		session relayer.SessionTree,
-	) (_ either.SessionTree, skip bool) {
+		sessionTrees []relayer.SessionTree,
+	) (_ either.SessionTrees, skip bool) {
 		rs.pendingTxMu.Lock()
 		defer rs.pendingTxMu.Unlock()
 
-		logger := polylog.Ctx(ctx)
-
-		// this session should no longer be updated
-		claimRoot, err := session.Flush()
-		if err != nil {
-			return either.Error[relayer.SessionTree](err), false
+		if len(sessionTrees) == 0 {
+			return either.Success(sessionTrees), false
 		}
 
-		// TODO_BLOCKER(@bryanchriswhite): The block that'll be used as a source of entropy for
-		// which branch(es) to prove should be deterministic and use on-chain governance params.
-		pathBlockHeight := session.GetSessionHeader().GetSessionEndBlockHeight() +
-			sessionkeeper.GetSessionGracePeriodBlockCount()
-		pathBlock, err := rs.blockQueryClient.Block(ctx, &pathBlockHeight)
-		if err != nil {
-			return either.Error[relayer.SessionTree](err), false
+		sessionClaims := []*relayer.SessionClaim{}
+		for _, session := range sessionTrees {
+			sessionClaims = append(sessionClaims, &relayer.SessionClaim{
+				RootHash:      session.GetClaimRoot(),
+				SessionHeader: session.GetSessionHeader(),
+			})
 		}
-		logger.Info().
-			Int64("session_start_block", pathBlock.Block.Height).
-			Msg("submitting claim")
-
-		sessionHeader := session.GetSessionHeader()
-		if err := rs.supplierClient.CreateClaim(ctx, *sessionHeader, claimRoot); err != nil {
-			failedCreateClaimSessionsPublishCh <- session
-			return either.Error[relayer.SessionTree](err), false
+
+		// Since all sessionTrees in the batch have the same start height and the
+		// slice is guaranteed to be non-empty, the first sessionTree's start height
+		// is used to log the claims submission.
+		sessionStartHeight := sessionTrees[0].GetSessionHeader().GetSessionStartBlockHeight()
```

Comment:
Same as https://github.com/pokt-network/poktroll/pull/521#discussion_r1596707068, how is this constraint ensured?